### PR TITLE
Use CMake standard library to handle CUDA

### PR DESCRIPTION
**Pull request by leezu**
## Description ##
Use CMake standard library to handle CUDA.
Makes our code more maintainable and may fix bugs, because we rely on code tested by all CMake users, instead of the small subset of MXNet CMake users.

Include FindCUDAToolkit that replaces the deprecated FindCUDA module as of CMake 1.17.
https://gitlab.kitware.com/cmake/cmake/merge_requests/4093

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [X] Refactor cuda handling in cmake build 

## Comments ##
- Bump CMake minimum version requirement due to
  https://gitlab.kitware.com/cmake/cmake/issues/19199
- Lazy consensus for the minimum version bump: https://lists.apache.org/thread.html/82ead692d172918de3afdeb27dbd533d37e1c084aa5738c5e353bbfd%40%3Cdev.mxnet.apache.org%3E

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -620,61 +594,17 @@ if(MSVC)
...
-            # Assuming building in a container that doesn't have CUDA installed (ie CPU-only build machine)
-            # so use the stub cuda driver shared library
-            if(EXISTS ${CUDA_LIBRARY_PATH}/stubs/libcuda.so)
-            link_directories(${CUDA_LIBRARY_PATH}/stubs)
-            endif()
-        endif()
-        endif()
-    endif()
- endif()
+  CUDA_SELECT_NVCC_ARCH_FLAGS(NVCC_FLAGS_ARCH Auto)
```
Comment (yajiedesign) [inline CMakeLists.txt]: CUDA_SELECT_NVCC_ARCH_FLAGS look like deprecated

**Comment by yajiedesign** [comment]
Consider integration https://github.com/apache/incubator-mxnet/pull/16980 and update cmake to 3.15.5.

File: CMakeLists.txt (hunk 1 above)
Comment (leezu) [reply CMakeLists.txt]: See https://gitlab.kitware.com/cmake/cmake/issues/19199
`CUDA_SELECT_NVCC_ARCH_FLAGS` is currently part of a deprecated module, but the command itself is going to stay around. They'll just move it to another module.

**Comment by leezu** [comment]
I'll integrate the changes of #16980 once it's merged. Or would you like me to do it directly?
I'm currently not sure why edge build is failing. So #16980 may be merged first.

**Comment by yajiedesign** [comment]
Currently #16980 is waiting for ci to upgrade cmake. I don't know when the upgrade is complete.

**Comment by yajiedesign** [comment]
i see https://github.com/apache/incubator-mxnet-ci/pull/17 is merged.When can upgrade  complete? I'm not sure about this process.

File: CMakeLists.txt (hunk 1 above)
Comment (yajiedesign) [reply CMakeLists.txt]: ok

**Comment by leezu** [comment]
@yajiedesign I think the change will be picked up within 7 days. But @marcoabreu may provide more details (or can force the change to be picked up earlier).

The current PR is also waiting for apache/incubator-mxnet-ci#17 to be picked up.

File: 3rdparty/mshadow/CMakeLists.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 2.8.7)
+cmake_minimum_required(VERSION 3.0.2)
```
Comment (larroy) [inline 3rdparty/mshadow/CMakeLists.txt]: For cuda nothing less than 3.12 seems to work in ubuntu 18.04

**Comment by larroy** [comment]
> i see [apache/incubator-mxnet-ci#17](https://github.com/apache/incubator-mxnet-ci/pull/17) is merged.When can upgrade complete? I'm not sure about this process.

We would need to upgrade the cmake in the container, anybody can send a PR

**Comment by leezu** [comment]
> > i see [apache/incubator-mxnet-ci#17](https://github.com/apache/incubator-mxnet-ci/pull/17) is merged.When can upgrade complete? I'm not sure about this process.
> 
> The change should be picked up as soon as you merge this from master in your branch.

What do you mean? It's for the Windows images used by the CI. When will the CI use the updated images?

File: 3rdparty/mshadow/CMakeLists.txt (hunk 2 above)
Comment (leezu) [reply 3rdparty/mshadow/CMakeLists.txt]: This file only adds the lint target. Let's just keep it unchanged for now thanks to 

> CMake keeps policy settings on a stack, so changes made by the cmake_policy command affect only the top of the stack. A new entry on the policy stack is managed automatically for each subdirectory to protect its parents and siblings.

**Comment by leezu** [comment]
@marcoabreu can you share the date when mxnet-CI will use the updated Windows AMI?

**Comment by larroy** [comment]
> > > i see [apache/incubator-mxnet-ci#17](https://github.com/apache/incubator-mxnet-ci/pull/17) is merged.When can upgrade complete? I'm not sure about this process.
> > 
> > 
> > The change should be picked up as soon as you merge this from master in your branch.
> 
> What do you mean? It's for the Windows images used by the CI. When will the CI use the updated images?

Oh yes I got confused sorry. We would have to update the AMIs used by CI. This needs to be done by an Amazon engineer. Marco is not involved.

**Comment by leezu** [comment]
Ok, thank you. I saw Marco mentioned this is done automatically every 7 days (https://github.com/apache/incubator-mxnet-ci/blob/master/services/jenkins-slave-creation-windows/README.md). Is it wrong?
@larroy could you help to get the update done? Thank you

**Comment by larroy** [comment]
I passed on your request to the engineering team.

**Comment by larroy** [comment]
Why are the dockcross images related to CMake changes??

**Comment by leezu** [comment]
Because the updated dockcross ships with cmake 13.

**Comment by leezu** [comment]
Rebased after https://github.com/apache/incubator-mxnet/pull/17098 got merged.

**Comment by leezu** [comment]
Ping for review

@szha @yajiedesign or anyone willing

File: cmake/Modules/FindCUDAToolkit.cmake [hunk 3]
```diff
@@ -0,0 +1,833 @@
...
+    set(CUDAToolkit_VERSION_MAJOR "${CMAKE_MATCH_1}")
+    set(CUDAToolkit_VERSION_MINOR "${CMAKE_MATCH_2}")
+    set(CUDAToolkit_VERSION_PATCH "${CMAKE_MATCH_3}")
+    set(CUDAToolkit_VERSION  "${CMAKE_MATCH_1}.${CMAKE_MATCH_2}.${CMAKE_MATCH_3}")
+  endif()
+  unset(NVCC_OUT)
+endif()
+
+
+get_filename_component(CUDAToolkit_ROOT_DIR ${CUDAToolkit_BIN_DIR} DIRECTORY ABSOLUTE)
```
Comment (ChaiBapchya) [inline cmake/Modules/FindCUDAToolkit.cmake]: I understand this has been included from https://gitlab.kitware.com/cmake/cmake/merge_requests/4093
However, I'm confused between `CUDAToolkit_ROOT_DIR` and `CUDA_TOOLKIT_ROOT_DIR` that's used at rest of the places in cmake in our codebase.
Are they both pointing to same thing? @leezu Thanks

Comment (leezu) [reply cmake/Modules/FindCUDAToolkit.cmake]: Yes, it should be insensitive to capitalization

**Pull request merged**